From 6a0f8c1ac98b7f773ea8b8e386b8e3bac77bdc63 Mon Sep 17 00:00:00 2001
From: Thomas Wood <thos@gnome.org>
Date: Sat, 29 Jan 2011 18:09:58 +0000
Subject: datetime: use a single policy kit action for configuring date and time

A single action allows the user to authenticate only once to change any
related date and time settings.
---
Index: gnome-settings-daemon-2.32.1/plugins/datetime/gsd-datetime-mechanism.c
===================================================================
--- gnome-settings-daemon-2.32.1.orig/plugins/datetime/gsd-datetime-mechanism.c	2011-02-18 09:13:51.843260254 -0500
+++ gnome-settings-daemon-2.32.1/plugins/datetime/gsd-datetime-mechanism.c	2011-02-18 09:32:20.518119349 -0500
@@ -223,8 +223,9 @@
 }
 
 static gboolean
-_check_polkit_for_action (GsdDatetimeMechanism *mechanism, DBusGMethodInvocation *context, const char *action)
+_check_polkit_for_action (GsdDatetimeMechanism *mechanism, DBusGMethodInvocation *context)
 {
+        const char *action = "org.gnome.settingsdaemon.datetimemechanism.configure";
         const char *sender;
         GError *error;
         PolkitSubject *subject;
@@ -275,7 +276,7 @@
 {
         GError *error;
 
-        if (!_check_polkit_for_action (mechanism, context, "org.gnome.settingsdaemon.datetimemechanism.settime"))
+        if (!_check_polkit_for_action (mechanism, context))
                 return FALSE;
 
         if (settimeofday (tv, NULL) != 0) {
@@ -433,7 +434,7 @@
         reset_killtimer ();
         g_debug ("SetTimezone('%s') called", zone_file);
 
-        if (!_check_polkit_for_action (mechanism, context, "org.gnome.settingsdaemon.datetimemechanism.settimezone"))
+        if (!_check_polkit_for_action (mechanism, context))
                 return FALSE;
 
         error = NULL;
@@ -541,8 +542,7 @@
 
         error = NULL;
 
-        if (!_check_polkit_for_action (mechanism, context,
-                                       "org.gnome.settingsdaemon.datetimemechanism.configurehwclock"))
+        if (!_check_polkit_for_action (mechanism, context))
                 return FALSE;
 
         if (g_file_test ("/sbin/hwclock", 
@@ -626,7 +626,7 @@
 
         error = NULL;
 
-        if (!_check_polkit_for_action (mechanism, context, "org.gnome.clockapplet.mechanism.configurentp"))
+        if (!_check_polkit_for_action (mechanism, context))
                 return FALSE;
 
         cmd = g_strconcat ("/usr/sbin/update-rc.d ntp ", using_ntp ? "enable" : "disable", NULL);
Index: gnome-settings-daemon-2.32.1/plugins/datetime/org.gnome.settingsdaemon.datetimemechanism.policy.in
===================================================================
--- gnome-settings-daemon-2.32.1.orig/plugins/datetime/org.gnome.settingsdaemon.datetimemechanism.policy.in	2011-02-18 09:13:51.813259765 -0500
+++ gnome-settings-daemon-2.32.1/plugins/datetime/org.gnome.settingsdaemon.datetimemechanism.policy.in	2011-02-18 09:32:20.518119349 -0500
@@ -8,40 +8,14 @@
   <vendor_url>http://www.gnome.org/</vendor_url>
   <icon_name>gnome-panel-clock</icon_name>
 
-  <action id="org.gnome.settingsdaemon.datetimemechanism.settimezone">
-    <_description>Change system time zone</_description>
-    <_message>Privileges are required to change the system time zone.</_message>
-    <defaults>
-      <allow_inactive>no</allow_inactive>
-      <allow_active>auth_self_keep</allow_active>
-    </defaults>
-  </action>
-
-  <action id="org.gnome.settingsdaemon.datetimemechanism.settime">
-    <_description>Change system time</_description>
-    <_message>Privileges are required to change the system time.</_message>
-    <defaults>
-      <allow_inactive>no</allow_inactive>
-      <allow_active>auth_admin_keep</allow_active>
-    </defaults>
-  </action>
-
-  <action id="org.gnome.settingsdaemon.datetimemechanism.configurehwclock">
-    <_description>Configure hardware clock</_description>
-    <_message>Privileges are required to configure the hardware clock.</_message>
+  <action id="org.gnome.settingsdaemon.datetimemechanism.configure">
+    <_description>Change system time and date settings</_description>
+    <_message>Privileges are required to configure time and date settings.</_message>
     <defaults>
+      <allow_any>no</allow_any>
       <allow_inactive>no</allow_inactive>
       <allow_active>auth_admin_keep</allow_active>
     </defaults>
   </action>
 
-  <action id="org.gnome.settingsdaemon.datetimemechanism.configurentp">
-    <_description>Configure network time</_description>
-    <_message>Privileges are required to configure network time.</_message>
-    <defaults>
-      <allow_inactive>no</allow_inactive>
-      <allow_active>auth_self_keep</allow_active>
-    </defaults>
-  </action>
-
 </policyconfig>
